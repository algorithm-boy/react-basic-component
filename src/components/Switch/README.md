# Switch

## API

| 参数(props) | 说明(specification) | 类型(type) | 默认值(defaultProps) |
| --- | --- | --- | --- |
| className(optional) | Switch的类名 (The class property for Switch) | string | - |
| style(optional) | Switch的行内样式 (The inline style for Switch) | object | {} |
| autoFocus(optional) | 自动聚焦 (Auto focus when componentDidMount) | boolean | - |
| prefixCls(optional) | 类名前缀 (The prefix of class) | string | 'hfdc-switch' |
| checked(optional) | 是否 `checked`，如果传入该值，则 `Switch` 组件为受控组件 (determine whether the `Switch` component is checked) | boolean | - |
| defaultChecked(optional) | `checked` 的初始值 (The `checked` inital state) | boolean | false |
| renderChecked(optional) | 状态是 `checked` 时展示的内容 (The content to be shown when the state is `checked`) | string \| ReactNode | - |
| renderUnChecked(optional) | 状态是 `unchecked` 时展示的内容 (The content to be shown when the state is `unchecked`) | string \| ReactNode | - |
| onChange(optional) | `checked` 状态更改时的回调 (Trigger when `checked` state changed) | Function(checked: boolean, event?: Event) | - |
| onClick(optional) | 点击的回调 (Trigger when clicked) | Function(checked: boolean, event: Event) | - |
| onMouseUp(optional) | 鼠标抬起时的回调 (Trigger when mouse up) | Function(checked: boolean, event: Event) | - |

## Example
```javascript
class SwitchDemo extends React.PureComponent<any, any> {
  constructor (props: any) {
    super(props);
    this.state = {
      checked: true
    }
  }
  render () {
    const { checked } = this.state;

    return <Switch
      renderChecked='开'
      renderUnChecked='关'
      prefixCls='diana-switch'
      checked={checked}
      onClick={(checked, e) => {
        console.log('triggle click', e)
        this.setState({ checked: !checked })
      }}
    />
  }
}
```