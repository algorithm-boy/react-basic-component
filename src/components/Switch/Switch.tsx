import React, { SFC, ReactNode, useState, useEffect, useRef } from 'react';
import { GeneralProps } from '../global.d';
import pick from 'lodash/pick';
import classnames from 'classnames';

export interface SwitchProps extends GeneralProps {
  autoFocus?: boolean;
  prefixCls?: string;
  checked?: boolean;
  defaultChecked?: boolean;
  disabled?: boolean;
  renderChecked?: string | ReactNode;
  renderUnChecked?: string | ReactNode;
  onChange?: (checked: boolean, event?: React.MouseEvent<HTMLButtonElement, MouseEvent> | React.KeyboardEvent<HTMLButtonElement>) => any;
  onClick?: (checked: boolean, event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => any;
  onMouseUp?: (checked: boolean, event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => any;
};

type DomProps = 'className' | 'style' | 'disabled';

const Switch: SFC<SwitchProps> = props => {
  const {
    autoFocus,
    prefixCls = 'hfdc-switch',
    className,
    checked: checkedProps,
    defaultChecked = false,
    disabled,
    renderChecked,
    renderUnChecked,
    onChange,
    onClick,
    onMouseUp
  } = props;
  const domProps: DomProps[] = ['style', 'disabled'];
  const pickedProps = pick(props, domProps);

  const button = useRef<HTMLButtonElement>(null);
  const [checked, setChecked] = useState<boolean>(checkedProps !== undefined ? checkedProps : defaultChecked);
  useEffect(() => {
    if (autoFocus && !disabled) {
      focus();
    }
  }, [autoFocus, disabled]);
  /* eslint-disable */
  // ignore eslint-plugin-react-hooks which will auto fill conditional variable
  useEffect(() => {
    handleChecked(undefined, defaultChecked);
  }, [checkedProps]);
  /* eslint-enable */

  const classes = classnames({
    [prefixCls]: true,
    [`${prefixCls}-checked`]: checked,
    [`${prefixCls}-disabled`]: disabled,
    [className || '']: !!className
  });

  function handleChecked(e?: React.MouseEvent<HTMLButtonElement, MouseEvent> | React.KeyboardEvent<HTMLButtonElement>, checkVal?: boolean) {
    if (disabled) return;
    if ('checked' in props) {
      if (checked === checkedProps) return;
      setChecked(checkedProps!);
      onChange && onChange(checkedProps!, e);
    } else {
      if (checked === checkVal) return;
      const val = checkVal !== undefined ? checkVal : !checked;
      setChecked(val);
      onChange && onChange(val, e);
    }
  }

  function handleClick(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
    onClick && onClick(checked, e);
    handleChecked(e);
  }

  function handleKeyDown(e: React.KeyboardEvent<HTMLButtonElement>) {
    if (e.keyCode === 37) {
      handleChecked(e, false);
    } else if (e.keyCode === 39) {
      handleChecked(e, true);
    }
  }

  function handleMouseUp(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
    blur();
    onMouseUp && onMouseUp(checked, e);
  }

  function focus() {
    button.current && button.current.focus();
  }

  function blur() {
    button.current && button.current.blur();
  }

  return (
    <>
      <button
        {...pickedProps}
        ref={button}
        type='button'
        role='switch'
        className={classes}
        aria-checked={checked}
        disabled={disabled}
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        onMouseUp={handleMouseUp}
      >
        <span className={`${prefixCls}-inner`}>
          {checked ? renderChecked : renderUnChecked}
        </span>
      </button>
    </>
  );
};

export default Switch;
