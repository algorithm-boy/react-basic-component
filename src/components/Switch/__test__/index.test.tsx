import * as React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Switch from '../index';

configure({ adapter: new Adapter() });

describe('Switch - renders', () => {
  it('renders correctly - shallow', () => {
    const wrapper = shallow(
      <Switch />
    );
    expect(wrapper.find('button').exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - render', () => {
    const wrapper = render(
      <Switch
        renderChecked='checked'
        renderUnChecked='unchecked'
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - mount', () => {
    const wrapper = mount(
      <div>
        <Switch
          renderChecked='checked'
          renderUnChecked='unchecked'
        />
      </div>
    );
    expect(wrapper.contains(<Switch
      renderChecked='checked'
      renderUnChecked='unchecked'
    />)).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Switch - parameters', () => {
  it('parameters check', () => {
    const wrapper = mount(<Switch
      renderChecked='开'
      renderUnChecked='关'
      prefixCls='rc-switch'
      autoFocus
      disabled
      defaultChecked
      checked={false}
    />);

    expect(wrapper.props().autoFocus).toEqual(true);
    expect(wrapper.props().disabled).toEqual(true);
    expect(wrapper.props().defaultChecked).toEqual(true);
    expect(wrapper.props().checked).toEqual(false);
    expect(wrapper.props().renderChecked).toEqual('开');
    expect(wrapper.props().renderUnChecked).toEqual('关');
    expect(wrapper.props().prefixCls).toEqual('rc-switch');
  });
});

describe('Switch - callbacks', () => {
  it('callback check - onClick', done => {
    const wrapper = mount(<Switch
      renderChecked='开'
      renderUnChecked='关'
      onClick={(checked, e) => {
        expect(checked).toEqual(false);
        expect(e).toBeTruthy();
        done();
      }}
    />);

    wrapper.simulate('click');
  });

  it('callback check - onClick_checked', done => {
    const wrapper = mount(<Switch
      renderChecked='开1'
      renderUnChecked='关1'
      checked={true}
      onClick={(checked, e) => {
        expect(checked).toEqual(true);
        expect(e).toBeTruthy();
        done();
      }}
    />);

    wrapper.simulate('click');
  });

  it('callback check - onChange_unchecked', done => {
    const wrapper = mount(<Switch
      renderChecked='开2'
      renderUnChecked='关2'
      onChange={(checked, e) => {
        expect(checked).toEqual(true);
        done();
      }}
    />);

    wrapper.simulate('click');
  });

  it('callback check - onChange_checked', done => {
    const wrapper = mount(<Switch
      renderChecked='开3'
      renderUnChecked='关3'
      defaultChecked
      onChange={(checked, e) => {
        expect(checked).toEqual(false);
        done();
      }}
    />);

    wrapper.simulate('click');
  });

  it('callback check - onChange_keyboard_关', done => {
    const wrapper = mount(<Switch
      renderChecked='开4'
      renderUnChecked='关4'
      defaultChecked
      onChange={(checked, e) => {
        expect(checked).toEqual(false);
        done();
      }}
    />);

    wrapper.simulate('keydown', {
      keyCode: 37
    });
  });

  it('callback check - onChange_keyboard_开', done => {
    const wrapper = mount(<Switch
      renderChecked='开5'
      renderUnChecked='关5'
      onChange={(checked, e) => {
        expect(checked).toEqual(true);
        done();
      }}
    />);

    wrapper.simulate('keydown', {
      keyCode: 39
    });
  });

  it('callback check - onMouseUp', done => {
    const wrapper = mount(<Switch
      renderChecked='开6'
      renderUnChecked='关6'
      onMouseUp={(checked, e) => {
        expect(checked).toEqual(false);
        expect(e).toBeTruthy();
        done();
      }}
    />);

    wrapper.simulate('mouseup');
  });
});