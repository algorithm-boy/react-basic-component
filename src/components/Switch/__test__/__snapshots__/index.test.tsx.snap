// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch - renders renders correctly - mount 1`] = `ReactWrapper {}`;

exports[`Switch - renders renders correctly - render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "aria-checked": "false",
      "class": "hfdc-switch",
      "role": "switch",
      "type": "button",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "hfdc-switch-inner",
        },
        "children": Array [
          Object {
            "data": "unchecked",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-checked": undefined,
      "class": undefined,
      "role": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-checked": undefined,
      "class": undefined,
      "role": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Switch - renders renders correctly - shallow 1`] = `ShallowWrapper {}`;
